File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.java
import java.util.*;
class Solution {
    public int solution(String s) {
        int answer = 0;
        
        for(int i = 0; i < s.length(); i++) {
            Stack<Character> stack = new Stack<>();
            int rc = 0;
            for(int j = 0; j < s.length(); j++) {
                char c = s.charAt(j);
                if(c == '[' || c == '{' || c=='(') {
                    stack.push(c);
                }
                else {
                    rc++;
                }
                if(!stack.isEmpty() && c == ']' && stack.peek() =='[') {
                    stack.pop();
                    rc--;
                }
                if(!stack.isEmpty() && c == '}' && stack.peek() == '{') {
                    stack.pop();
                    rc--;
                }
                if(!stack.isEmpty() && c ==')' && stack.peek() == '(') {
                    stack.pop();
                    rc--;
                }
            }
            if(stack.isEmpty() && rc == 0) answer++;
            
            s = s.substring(1) + s.substring(0,1);
        }
        return answer;
    }
}

File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.js
function solution(s) {
    var answer = 0;
    
    function isRight(str) {
        const stack = [];
        for (const ch of str) {
            if (ch === '(' || ch === '{' || ch === '[') {
                stack.push(ch);
            } else {
                if (stack.length === 0) return false;
                const top = stack[stack.length-1];
                if (ch ===')' && top ==='(') stack.pop();
                else if (ch ==='}' && top === '{') stack.pop();
                else if (ch === ']' && top === '[') stack.pop();
            }
            
        }
        
        return stack.length === 0;
    }
    
    for (let i = 0; i < s.length; i++) {
        const rotated = s.slice(i) + s.slice(0,i);
        if (isRight(rotated)) answer++;
    }
    return answer;
}